block/bpf_io_filter.c

```diff
@@ -121,52 +128,108 @@ int io_filter_prog_attach(const union bpf_attr *attr, struct bpf_prog *prog)
 	if (disk == NULL)
 		return -ENXIO;
 
-	//TODO: change to array of programs to allow multiple programs to attach
-	rcu_assign_pointer(disk->prog, prog);
+	ret = mutex_lock_interruptible(&disk->io_filter_lock);
+	if (ret)
+		return ret;
+
+	old_array = io_filter_rcu_dereference_progs(disk);
+	if (old_array && bpf_prog_array_length(old_array) >= BPF_MAX_PROGS) {
+		ret = -E2BIG;
+		goto unlock;
+	}
+
+	ret = bpf_prog_array_copy(old_array, NULL, prog, &new_array);
+	if (ret < 0)
+		goto unlock;
+
+	rcu_assign_pointer(disk->progs, new_array);
+	bpf_prog_array_free(old_array);
```

Comment:
Hmm yes it seems you should generally call synchronize_rcu() before freeing the old element. I followed the lirc example because it also uses a program array, but it does not call synchonize_rcu()  (https://elixir.bootlin.com/linux/latest/source/drivers/media/rc/bpf-lirc.c#L135). I am wondering if this is because the bpf_prog_array_free() has this functionality built in. The bpf_prog_array_free() uses kfree_rcu() which seems to wait a grace period. It could also have something to do with the way bpf_prog_array_copy is implemented. Let me know if you think the synchronize_rcu call is still necessary

Reply:
Okay sounds good, looks like bpf_prog_array_free() takes care of that. I also saw some more examples and they don't call syncrhonize_rcu() after bpf_prog_array_free() so I think we are good.